src/hid-asus.c

```diff
@@ -0,0 +1,320 @@
+		int id = input_mt_get_value(ps, ABS_MT_TRACKING_ID);
+
+		if (id < 0)
+			continue;
+		if ((id - oldid) & TRKID_SGN) {
+			oldest = ps;
+			oldid = id;
+		}
+		count++;
+	}
+
+	if (oldest) {
+		input_report_abs(input, ABS_TOOL_WIDTH,
+			input_mt_get_value(oldest, ABS_MT_TOUCH_MAJOR));
+	}
+}
+
+static void asus_report_input(struct input_dev *input, u8 *data)
+{
+	int i;
+	u8 *contactData = data + CONTACT_DATA_OFFSET;
+
+	for (i = 0; i < MAX_CONTACTS; i++) {
+		bool down = data[CONTACT_DOWN_OFFSET] &
+				(CONTACT_DOWN_MASK << i);
+		int toolType = contactData[CONTACT_TOOL_TYPE_OFFSET] &
+				CONTACT_TOOL_TYPE_MASK ? MT_TOOL_PALM
+				: MT_TOOL_FINGER;
+
+		input_mt_slot(input, i);
+		input_mt_report_slot_state(input, toolType, down);
+
+		if (down) {
+			asus_report_contact_down(input, toolType, contactData);
+			contactData += CONTACT_DATA_SIZE;
+		}
+	}
+
+	input_report_key(input, BTN_LEFT,
+			data[BTN_LEFT_OFFSET] & BTN_LEFT_MASK);
+	asus_report_tool_width(input);
+
+	input_mt_sync_frame(input);
+	input_sync(input);
+}
+
+static int asus_raw_event(struct hid_device *hdev,
+		struct hid_report *report, u8 *data, int size)
+{
+	if (hdev->product == USB_DEVICE_ID_ASUSTEK_TOUCHPAD &&
+			 data[REPORT_ID_OFFSET] == INPUT_REPORT_ID &&
+						size == INPUT_REPORT_SIZE) {
+		struct hid_input *hidinput;
+
+		list_for_each_entry(hidinput, &hdev->inputs, list) {
+			asus_report_input(hidinput->input, data);
+		}
+		return 1;
+	}
+
+	return 0;
+}
+
+static int asus_input_configured(struct hid_device *hdev, struct hid_input *hi)
+{
+	if (hdev->product == USB_DEVICE_ID_ASUSTEK_TOUCHPAD) {
+		int ret;
+		struct input_dev *input = hi->input;
+
+		input->name = "Asus TouchPad";
+
+		input_set_abs_params(input, ABS_MT_POSITION_X, 0, MAX_X, 0, 0);
+		input_set_abs_params(input, ABS_MT_POSITION_Y, 0, MAX_Y, 0, 0);
+		input_set_abs_params(input, ABS_TOOL_WIDTH, 0, MAX_TOUCH_MAJOR, 0, 0);
+		input_set_abs_params(input, ABS_MT_TOUCH_MAJOR, 0, MAX_TOUCH_MAJOR, 0, 0);
+		input_set_abs_params(input, ABS_MT_PRESSURE, 0, MAX_PRESSURE, 0, 0);
+
+		__set_bit(BTN_LEFT, input->keybit);
+		__set_bit(INPUT_PROP_BUTTONPAD, input->propbit);
+
+		ret = input_mt_init_slots(input, MAX_CONTACTS, INPUT_MT_POINTER);
+
+		if (ret) {
+			hid_err(hdev, "Asus input mt init slots failed: %d\n", ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int asus_input_mapping(struct hid_device *hdev,
+		struct hid_input *hi, struct hid_field *field,
+		struct hid_usage *usage, unsigned long **bit,
+		int *max)
+{
+	if (hdev->product == USB_DEVICE_ID_ASUSTEK_TOUCHPAD) {
+		/* Don't map anything from the HID report.
+		 * We do it all manually in asus_input_configured
+		 */
+		return -1;
+	}
+
+	return 0;
+}
+
+static int asus_start_multitouch(struct hid_device *hdev)
+{
+	unsigned char buf[] = { FEATURE_REPORT_ID, 0x00, 0x03, 0x01, 0x00 };
+	int ret = hid_hw_raw_request(hdev, FEATURE_REPORT_ID, buf, sizeof(buf),
+			HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
```

Comment:
You can't call hid_hw_raw_request with non-DMA capable buffers. (see https://patchwork.kernel.org/patch/9439221/)

It might not be an issue for I2C devices, but it's one for USB.